mlir/lib/CAPI/IR/IR.cpp

```diff
@@ -1059,6 +1059,26 @@ void mlirBlockPrint(MlirBlock block, MlirStringCallback callback,
   unwrap(block)->print(stream);
 }
 
+intptr_t mlirBlockGetNumSuccessors(MlirBlock block) {
+  return static_cast<intptr_t>(unwrap(block)->getNumSuccessors());
+}
+
+MlirBlock mlirBlockGetSuccessor(MlirBlock block, intptr_t pos) {
+  return wrap(unwrap(block)->getSuccessor(static_cast<unsigned>(pos)));
+}
+
+intptr_t mlirBlockGetNumPredecessors(MlirBlock block) {
+  Block *b = unwrap(block);
+  return static_cast<intptr_t>(std::distance(b->pred_begin(), b->pred_end()));
+}
+
+MlirBlock mlirBlockGetPredecessor(MlirBlock block, intptr_t pos) {
+  Block *b = unwrap(block);
+  Block::pred_iterator it = b->pred_begin();
+  std::advance(it, pos);
+  return wrap(*it);
```

Comment:
I'd rather avoid iterating over the use-def list every time... This goes through block's use-def chain, maybe there is a way to expose a `BlockOperand` (and incidentally `OpOperand` if it isn't) and a `getNextUse`.

Reply:
Don't think so

https://github.com/llvm/llvm-project/blob/1b5d6ec6855369d109fcb740ecd3812231b7a279/mlir/include/mlir/IR/BlockSupport.h#L45

Compare with `SuccessorRange` just below there.

Reply:
Define `predecessor` as returning an iterable object that only has `__next__`?

Not a big problem if we don't, we use indexed accessors for the linked list of blocks as well because I thought it whoever was using Python didn't care about that level of performance tweaking.

Reply:
The problem is there's no way to implement `GetNextPredecessor` like `GetNextBlockInRegion` 

https://github.com/llvm/llvm-project/blob/main/mlir/lib/CAPI/IR/IR.cpp#L969

without holding an instance of `PredecessorIterator` (even forgetting that it needs to be mapped into C). 

Anyway ya I'm gonna have this as is but I'll add a comment mentioning that it's expensive.